nsqdserver/protocol_v2.go

```diff
@@ -1395,11 +1425,39 @@ func (p *protocolV2) internalPubAndTrace(client *nsqd.ClientV2, params [][]byte,
 	messageBody := messageBodyBuffer.Bytes()[:bodyLen]
 
 	partition := topic.GetTopicPart()
+	var extJsonLen uint16
 	var traceID uint64
+	var existInJsonHeader bool
 	var realBody []byte
-	if traceEnable {
+	if traceEnable && !pubExt {
 		traceID = binary.BigEndian.Uint64(messageBody[:nsqd.MsgTraceIDLength])
 		realBody = messageBody[nsqd.MsgTraceIDLength:]
+	} else if pubExt {
+		//read two byte header length
+		extJsonLen = binary.BigEndian.Uint16(messageBody[:nsqd.MsgJsonHeaderLength])
+		extJsonBytes := messageBody[nsqd.MsgJsonHeaderLength:nsqd.MsgJsonHeaderLength + extJsonLen]
+		jhe, ok := extContent.(*ext.JsonHeaderExt)
+		if !ok {
+			return nil, fmt.Errorf("invalid ext content type, Json Header Ext expected, got %v", reflect.TypeOf(extContent))
+		}
+		//validate json header passin
+		jsonHeader, err := simpleJson.NewJson(extJsonBytes)
+		if err != nil {
+			return nil, protocol.NewClientErr(err, ext.E_INVALID_JSON_HEADER, "fail to parse json header")
+		}
+
+		//parse traceID, if there is any
+		traceIDJson := jsonHeader.Get(ext.TRACE_ID_KEY)
+		traceIDJson, existInJsonHeader = jsonHeader.CheckGet(ext.TRACE_ID_KEY)
+		if existInJsonHeader {
+			traceID, err = traceIDJson.Uint64()
```

Comment:
Maybe add var `needTracedRsp` and should set to true if we found valid traceid here